bin/build

```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
+
+appDir="$(pwd)"
+bpDir="$(cd $(dirname $0)/..; pwd)" # absolute path
+layersDir="${1:?}"
+
+source "${bpDir}/lib/v3/common.sh"
+source "${bpDir}/lib/v3/buildpack.sh"
+source "${bpDir}/lib/jvm.sh"
+
+status "Installing Java"
+
+# fail hard
+set -o pipefail
+# fail harder
+set -eu
+
+# TODO check the build plan to see if another JDK has already been installed?
+
+bp_install_or_reuse_toolbox "${layersDir}/utils"
+
+jdkVersion="$(get_jdk_version "${appDir}")"
+
+jdkUrl="$(get_jdk_url "${jdkVersion}")"
+
+jdkCacheId="$(get_jdk_cache_id "${jdkUrl}")"
+
+jdkMetadata=$(cat <<EOF
+version = "${jdkVersion}"
+cache_id = "${jdkCacheId}"
+EOF
+)
+
+if bp_layer_has_key? "${layersDir}/jdk" ".metadata.cache_id" "${jdkCacheId}"; then
+  info "Reusing Java ${jdkVersion}"
+else
+  info "Installing JDK ${jdkVersion}"
+  jdkLayerToml="$(bp_layer_metadata_create "false" "true" "true" "${jdkMetadata}")"
+  jdkLayer="$(bp_layer_init "${layersDir}" "jdk" "${jdkLayerToml}")"
+
+  install_jdk "${jdkVersion}" "${jdkUrl}" "${jdkLayer}"
+  install_certs "${jdkLayer}"
+  install_profile "${bpDir}" "${jdkLayer}/profile.d"
+  install_jdk_overlay "${jdkVersion}" "${jdkLayer}" "${appDir}"
+
+  info "Installing JRE ${jdkVersion}"
+  jreLayerToml="$(bp_layer_metadata_create "true" "false" "false" "${jdkMetadata}")"
+  jreLayer="$(bp_layer_init "${layersDir}" "jre" "${jreLayerToml}")"
+  install_jre "${jdkLayer}" "${jreLayer}"
```

Comment:
should this JRE layer logic be treated and checked separately?

Reply:
I don't think it's necessary because the JRE is contained within the JDK (so it's essentially extracted from that layer). They are very much entangled.

Reply:
Should the JRE layer contain the ETAG from the JRE tarball?

Reply:
there is no JRE tarball. it's a subdir of the exploded JDK